src/LLVM_intrinsics.jl

```diff
@@ -0,0 +1,530 @@
+# LLVM operations and intrinsics
+module LLVM
+
+# TODO: fastmath flags
+
+import ..SIMD: VE, LVec, IntegerTypes, IntTypes, UIntTypes, FloatingTypes, ScalarTypes
+
+const d = Dict{DataType, String}(
+    Bool         => "i8",
+    Int8         => "i8",
+    Int16        => "i16",
+    Int32        => "i32",
+    Int64        => "i64",
+    Int128       => "i128",
+
+    UInt8        => "i8",
+    UInt16       => "i16",
+    UInt32       => "i32",
+    UInt64       => "i64",
+    UInt128      => "i128",
+
+    #Float16     => "half",
+    Float32      => "float",
+    Float64      => "double",
+)
+# Add the Ptr translations
+foreach(x -> (d[Ptr{x}] = d[Int]), collect(keys(d)))
+# d[Ptr{Nothing}] = d[Int]
+
+# LT = LLVM Type (scalar and vectors), we keep type names intentionally short
+# to make the signatures smaller
+const LT{T} = Union{LVec{<:Any, T}, T}
+
+
+# TODO: Clean up
+suffix(N::Integer, ::Type{Ptr{T}}) where {T} = "v$(N)p0$(T<:IntegerTypes ? "i" : "f")$(8*sizeof(T))"
+suffix(N::Integer, ::Type{T}) where {T} = "v$(N)$(T<:IntegerTypes ? "i" : "f")$(8*sizeof(T))"
+suffix(::Type{T}) where {T} = "$(T<:IntegerTypes ? "i" : "f")$(8*sizeof(T))"
+llvm_name(llvmf, N, T) = string("llvm", ".", llvmf, ".", suffix(N, T))
+llvm_name(llvmf, ::Type{LVec{N, T}}) where {N,T} = string("llvm", ".", llvmf, ".", suffix(N, T))
+llvm_name(llvmf, ::Type{T}) where {T<:ScalarTypes} = string("llvm", ".", llvmf, ".", suffix(T))
+llvm_type(::Type{T}) where {T<:ScalarTypes} = d[T]
+llvm_type(::Type{LVec{N, T}}) where {N,T} = "< $N x $(d[T])>"
+
+####################
+# Unary operators  #
+####################
+
+const UNARY_INTRINSICS_FLOAT = [
+    :sqrt
+    :sin
+    :cos
+    :exp
+    :trunc
+    :exp2
+    :log
+    :log10
+    :log2
+    :fabs
+    :floor
+    :ceil
+    :rint
+    :nearbyint
+    :round
+]
+
+const UNARY_INTRINSICS_INT = [
+    :bitreverse
+    :bswap
+    :ctpop
+    :ctlz
+    :cttz
+    :fshl
+    :fshr
+]
+
+for (fs, c) in zip([UNARY_INTRINSICS_FLOAT, UNARY_INTRINSICS_INT], [FloatingTypes, IntegerTypes])
+    for f in fs
+        @eval begin
+            @generated function $(f)(x::T) where T<:LT{<:$c}
```

Comment:
Since you are already using `@eval` you can avoid the generated function here.
https://github.com/JuliaGPU/KernelAbstractions.jl/blob/50efa419a18947b9c277d235360be7e5919a8f6b/src/compiler/contract.jl#L14

Reply:
The type `T` is not yet known in the `@eval`. To avoid the `@generated`, there needs to be an extra loop over all types `T`.

Reply:
You also need to loop over `N` though?

Reply:
Ah I see that particular bit of magic behind `LT` slipped past me. Sometimes I whish LLVM intrinsic would actually have multiple dispatch instead of just encoding it in a string. In theory it is supposed to be possible to have LLVM infer the type of the intrinsic from the arguments. but alas I haven't figure out how to make the work from the C interface.
https://llvm.org/doxygen/namespacellvm_1_1Intrinsic.html